warmrestart/warmRestartAssist.cpp

```diff
@@ -79,7 +79,10 @@ void AppRestartAssist::registerAppTable(const std::string &tableName, ProducerSt
     m_psTables[tableName]  = psTable;
 
     // Clear the producerstate table to make sure no pending data for the AppTable
-    psTable->clear();
+    if (m_warmStartInProgress)
+    {
+        psTable->clear();
```

Comment:
Looks like this change will affect both `neighsyncd` and `natsyncd`. @lguohan is it OK to not clear the table for `natsyncd` when the dut is warm-rebooting? If not and to limit the change to limit `neighsyncd`, we can also check the table name to be `neighsyncd`

Reply:
Wouldn't nat hit same issue if this protection is not there?

Reply:
It is not clear if it is required for NAT tables to be cleared unconditionally or not. Needs a bit of digging into how NAT is using this shared class. NAT is using this method for 4 table [here](https://github.com/Azure/sonic-swss/blob/master/natsyncd/natsync.cpp#L57)

Reply:
This is a bug introduced by https://github.com/Azure/sonic-swss/pull/1126 where the refactoring of the library to support multiple tables was calling this psTable->clear unconditionally. The way the fix doing here should be correct.  NAT tables or any client should not use the library to flush producer state table in non warm-reboot cases.

Reply:
btw: The reason we use psTable clear was to make sure the relevant table wouldn’t change in some corner cases after we dumped it to memory,  this is required only in warm-reboot/restart case before we dump the table.  Also, since the daemon using the library was the producer itself, it was safe to do so.   However, this assumption was broken if we use swssconfig to load the table at the same time,  which is the case for non warm-reboot cases for arp, nat tables etc.  In those cases, the library cleared the requests from swssconfig incorrectly and cause the issues reported.